File: LowRes NX Coder/LowRes NX Coder/Scenes/Help/HelpTextViewController.m
//
//  HelpTextViewController.m
//  Pixels
//
//  Created by Timo Kloss on 25/12/14.
//  Copyright (c) 2014 Inutilis Software. All rights reserved.
//

#import "HelpTextViewController.h"
#import "HelpTableViewController.h"
#import "HelpContent.h"
#import "HelpSplitViewController.h"
#import "LowRes_NX_Coder-Swift.h"

@interface HelpTextViewController ()

@property (weak, nonatomic) IBOutlet UIWebView *webView;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *activityView;

@end

@implementation HelpTextViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.webView.delegate = self;
    self.webView.scalesPageToFit = NO;
    self.webView.dataDetectorTypes = UIDataDetectorTypeNone;
    
    self.navigationItem.leftBarButtonItem = self.splitViewController.displayModeButtonItem;
    self.navigationItem.leftItemsSupplementBackButton = YES;
    
    HelpContent *helpContent = AppController.shared.helpContent;
    [self.webView loadHTMLString:helpContent.manualHtml baseURL:helpContent.url];
}

- (void)setChapter:(NSString *)chapter
{
    _chapter = chapter;
    if (!self.webView.isLoading)
    {
        [self jumpToChapter:self.chapter];
    }
}

- (void)jumpToChapter:(NSString *)chapter
{
    NSString *script = [NSString stringWithFormat:@"document.getElementById('%@').scrollIntoView(true);", chapter];
    [self.webView stringByEvaluatingJavaScriptFromString:script];
}

- (BOOL)webView:(UIWebView *)webView shouldStartLoadWithRequest:(NSURLRequest *)request navigationType:(UIWebViewNavigationType)navigationType
{
    if (!request.URL.isFileURL)
    {
        // open links in Safari
        [[UIApplication sharedApplication] openURL:[request URL]];
        return NO;
    }
    return YES;
}

- (void)webViewDidStartLoad:(UIWebView *)webView
{
    [self.activityView startAnimating];
}

- (void)webViewDidFinishLoad:(UIWebView *)webView
{
    [self.activityView stopAnimating];
    if (self.chapter)
    {
        [self jumpToChapter:self.chapter];
    }
}

@end

File: LowRes NX Coder/LowRes NX Coder/Scenes/Help/HelpTextViewController.swift
//
//  HelpTextViewController.swift
//  LowRes NX Coder
//
//  Created by Timo Kloss on 12.03.21.
//  Copyright © 2021 Inutilis Software. All rights reserved.
//

import UIKit
import WebKit

class HelpTextViewController: UIViewController, WKNavigationDelegate {
    
    private var chapter: String?
    
    private var webView: WKWebView!
    @IBOutlet weak var activityView: UIActivityIndicatorView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let config = WKWebViewConfiguration()
        webView = WKWebView(frame: view.bounds, configuration: config)
        webView.translatesAutoresizingMaskIntoConstraints = false
        webView.navigationDelegate = self
        
        view.insertSubview(webView, at: 0)
        
        webView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor).isActive = true
        webView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor).isActive = true
        webView.leftAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leftAnchor).isActive = true
        webView.rightAnchor.constraint(equalTo: view.safeAreaLayoutGuide.rightAnchor).isActive = true
        
        navigationItem.leftBarButtonItem = splitViewController?.displayModeButtonItem
        navigationItem.leftItemsSupplementBackButton = true
        
        let helpContent = AppController.shared.helpContent
        webView.loadHTMLString(helpContent.manualHtml, baseURL: helpContent.url)
    }
    
    @objc func setChapter(_ chapter: String) {
        self.chapter = chapter
        if !webView.isLoading {
            jumpToChapter(chapter)
        }
    }
    
    private func jumpToChapter(_ chapter: String) {
        webView.evaluateJavaScript("document.getElementById('\(chapter)').scrollIntoView(true);") { (result, error) in
            if let error = error {
                print(error.localizedDescription)
            }
        }
    }
    
    func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
        activityView.startAnimating()
    }
    
    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        activityView.stopAnimating()
        if let chapter = chapter {
            jumpToChapter(chapter)
        }
    }
    
    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        activityView.stopAnimating()
    }
    
}
